addon-test-support/@ember/test-helpers/dom/focus.js

```diff
@@ -0,0 +1,56 @@
+import getElement from './-get-element';
+import fireEvent from './fire-event';
+import settled from '../settled';
+import isFocusable from './-is-focusable';
+import { nextTickPromise } from '../-utils';
+
+/**
+  @private
+  @method __focus__
+  @param {Element} element
+*/
+export function __focus__(element) {
+  let browserIsNotFocused = document.hasFocus && !document.hasFocus();
+
+  // makes `document.activeElement` be `element`. If the browser is focused, it also fires a focus event
+  element.focus();
+
+  // Firefox does not trigger the `focusin` event if the window
+  // does not have focus. If the document does not have focus then
+  // fire `focusin` event as well.
+  if (browserIsNotFocused) {
+    // if the browser is not focused the previous `el.focus()` didn't fire an event, so we simulate it
+    fireEvent(element, 'focus', {
+      bubbles: false,
+    });
+
+    fireEvent(element, 'focusin');
+  }
+}
+
+/**
+  @method focus
+  @param {String|Element} target
+  @return {Promise<void>}
+  @public
+*/
+export default function focus(target) {
+  if (!target) {
+    throw new Error('Must pass an element or selector to `focus`.');
+  }
+
+  let element = getElement(target);
+  if (!element) {
+    throw new Error(`Element not found when calling \`focus('${target}')\`.`);
+  }
+
+  if (!isFocusable(element)) {
+    throw new Error(`${target} is not focusable`);
+  }
```

Comment:
I think I raised this issue before but maybe it got lost, I'm a little worried about:

- throwing sync errors from functions that return a Promise. from what I've read about Promises so far, the best practice for async functions is to always reject instead of throwing sync errors.

- assertions about the input params are running before the actual focus code. as in, while the `nextTickPromise()` is running the element might be remove from the page which would result in an exception. maybe we should move those conditions into the `.then()` block 🤔

Reply:
Yes, you are correct. I experimented with how async functions work, and confirmed your logic. I will update to rejections before landing.